### PR TITLE
Many symintifications

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87722
* #87294
* __->__ #87604

Adds
expand_inplace
conv conv_double_backward
convolution
adaptive_avg_pool2d_symint
_embedding_bag_backward_symint
cudnn_grid_sampler
cuda 32 bit indexing
nll_loss / nll_loss_2d
tensor split
pooling same mode
cudnn_is_acceptable
storage nbytes

### CODE REVIEW

**Comment by albanD** [comment]
Will wait for CI for all the xfail update and make sure everything is tested.

File: aten/src/ATen/native/NonSymbolicBC.h
Comment (bdhirsh) [inline aten/src/ATen/native/NonSymbolicBC.h:25]: This has a small risk of internal breakage, right?

Although I just searched internal for `at::native::tensor_split`, and the only place that shows up is in static runtime (which is part of the OSS codebase). So.. I think this should be fine

File: aten/src/ATen/native/Pool.h
Comment (bdhirsh) [inline aten/src/ATen/native/Pool.h:81]: Do you think it's worth implementing `operator--` as part of merging to master? (I'll leave it up to you)

Comment (ezyang) [reply aten/src/ATen/native/Pool.h:81]: seems worthwhile. But I'll take a rain check if someone agrees to do ALL of the operator overloads, methodically

File: aten/src/ATen/native/TensorShape.cpp
Comment (bdhirsh) [inline aten/src/ATen/native/TensorShape.cpp:774]: hmmm don't we need to symintify that `self.size(dim_)`?

I guess we could add a template for `at::symint::size()`, and make it:
```
at::symint::size<T>(self)`
```
wdyt?

File: aten/src/ATen/native/native_functions.yaml
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:1824]: hmmm, why?

`CompositExplicitAutogradNonFunctional` is a bit funny today - it's basically only used for providing better error messages for XLA/LTC. If we have a functional op that calls a view op, that is `CompositeExplicitAutograd`, then XLA and LTC can't handle it directly so we don't want to register it to them. Not clear to me why we had to remove it for `diag_embed` here though.

File: aten/src/ATen/native/native_functions.yaml
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:7711]: Is there a comment explaining the reasoning for the change to `diag`?

It looks like before, `diag` was a `CompositeImplicitAutograd` op that called `diag_embed`, and now we have a brand new backward op for it.

File: aten/src/ATen/native/TensorShape.cpp
Comment (bdhirsh) [inline aten/src/ATen/native/TensorShape.cpp:3678]: This seems wrong - we're removing error checking?

File: aten/src/ATen/native/NonSymbolicBC.h
Comment (albanD) [reply aten/src/ATen/native/NonSymbolicBC.h:25]: Ok, let's take the risk on this one. Thanks for checking!

File: aten/src/ATen/native/TensorShape.cpp
Comment (albanD) [reply aten/src/ATen/native/TensorShape.cpp:774]: Yes I can add that.

File: aten/src/ATen/native/native_functions.yaml
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:1824]: Not sure tbh, I haven't reviewed this. Reverting.

File: torchgen/api/cpp.py
Comment (albanD) [reply torchgen/api/cpp.py:351]: It is actually used for `SymInt[3] padding=0` so it is for the list case.

File: aten/src/ATen/native/TensorShape.cpp
Comment (ezyang) [inline aten/src/ATen/native/TensorShape.cpp:3406]: The diff here is weirdly big and makes me nervous

**Comment by ezyang** [comment]
Everything here looks good except diag, which is weirdly complicated for what I expect

File: aten/src/ATen/native/native_functions.yaml
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:7711]: This is a mistake. The diag improvements were added to master indepently and never added to the branch.
I'll make sure not to revert https://github.com/pytorch/pytorch/pull/87180 by mistake :p

**Comment by albanD** [comment]
>  Everything here looks good except diag, which is weirdly complicated for what I expect

Yes the diag change is a mistake due to a land race with master. Cleaning this up now.

File: aten/src/ATen/native/Pool.h
Comment (albanD) [reply aten/src/ATen/native/Pool.h:81]: Ok, leaving all the operators for a follow up then.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**
